src/integrationTest/java/org/opensearch/security/api/InternalUsersRestApiIntegrationTest.java

```diff
@@ -0,0 +1,808 @@
+        withUser(ADMIN_USER_NAME, client -> settingOfUnknownRoleIsNotAllowed(userWithSecurityRoles, client));
+
+        withUser(
+            ADMIN_USER_NAME,
+            localCluster.getAdminCertificate(),
+            client -> settingOfUnknownRoleIsNotAllowed(userWithSecurityRoles, client)
+        );
+        withUser(ADMIN_USER_NAME, localCluster.getAdminCertificate(), this::canAssignedHiddenRole);
+
+        for (final var restAdminUser : List.of(REST_ADMIN_USER, REST_API_ADMIN_INTERNAL_USERS_ONLY)) {
+            withUser(restAdminUser, client -> settingOfUnknownRoleIsNotAllowed(userWithSecurityRoles, client));
+            withUser(restAdminUser, localCluster.getAdminCertificate(), this::canAssignedHiddenRole);
+        }
+    }
+
+    void impossibleToSetHiddenRoleIsNotAllowed(final String predefinedUserName, final TestRestClient client) throws Exception {
+        // put
+        notFound(
+            () -> client.putJson(
+                apiPath(randomAsciiAlphanumOfLength(10)),
+                internalUser(randomAsciiAlphanumOfLength(10), null, null, configJsonArray(HIDDEN_ROLE))
+            ),
+            "Resource 'hidden-role' is not available."
+        );
+        // patch
+        notFound(
+            () -> client.patch(
+                apiPath(),
+                patch(
+                    addOp(
+                        randomAsciiAlphanumOfLength(10),
+                        internalUser(randomAsciiAlphanumOfLength(10), null, null, configJsonArray(HIDDEN_ROLE))
+                    )
+                )
+            )
+        );
+        // TODO related to issue #4426
+        notFound(
+            () -> client.patch(apiPath(predefinedUserName), patch(addOp("opendistro_security_roles", configJsonArray(HIDDEN_ROLE))))
+
+        );
+    }
+
+    void canAssignedHiddenRole(final TestRestClient client) throws Exception {
+        final var userNamePut = randomAsciiAlphanumOfLength(4);
+        created(
+            () -> client.putJson(
+                apiPath(userNamePut),
+                internalUser(randomAsciiAlphanumOfLength(10), null, null, configJsonArray(HIDDEN_ROLE))
+            )
+        );
+    }
+
+    void settingOfUnknownRoleIsNotAllowed(final String predefinedUserName, final TestRestClient client) throws Exception {
+        notFound(
+            () -> client.putJson(
+                apiPath(randomAsciiAlphanumOfLength(10)),
+                internalUser(randomAsciiAlphanumOfLength(10), null, null, configJsonArray("unknown-role"))
+            ),
+            "role 'unknown-role' not found."
+        );
+        notFound(
+            () -> client.patch(
+                apiPath(),
+                patch(
+                    addOp(
+                        randomAsciiAlphanumOfLength(4),
+                        internalUser(randomAsciiAlphanumOfLength(10), null, null, configJsonArray("unknown-role"))
```

Comment:
What happens if you:

1. map a user to an existing role through `user.opendistro_security_roles`
2. Delete the role
3. Update the user

Will the update request go through or would it produce not found?

Reply:
Since we do not check data integrity:
- role will be deleted
- if you set new `opendistro_security_roles` it will be replaced
- if you do not add new `opendistro_security_roles` and change something else the operation will be succeeded.  Correct me if I misunderstand something.